Marlin/src/lcd/extui/lib/anycubic_i3mega/anycubic_i3mega_lcd.cpp

```diff
@@ -175,43 +175,48 @@ void AnycubicTFTClass::OnUserConfirmRequired(const char * const msg) {
-      mediaPrintingState = AMPRINTSTATE_PAUSED;
-      mediaPauseState    = AMPAUSESTATE_PARKING;
-      // TODO: JBA I don't think J05 just disables the continue button, i think it injects a rogue M25. So taking this out
-      // disable continue button
-      // SENDLINE_DBG_PGM("J05", "TFT Serial Debug: UserConfirm SD Filament Purging... J05"); // J05 printing pause
-
-      // enable continue button
-      SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm Filament is purging... J18");
-    }
-    else if (strcmp_P(msg, PSTR("HeaterTimeout")) == 0) {
-      mediaPrintingState = AMPRINTSTATE_PAUSED;
-      mediaPauseState    = AMPAUSESTATE_HEATER_TIMEOUT;
-      // enable continue button
-      SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm SD Heater timeout... J18");
-    }
-    else if (strcmp_P(msg, PSTR("Reheat finished.")) == 0) {
-      mediaPrintingState = AMPRINTSTATE_PAUSED;
-      mediaPauseState    = AMPAUSESTATE_REHEAT_FINISHED;
-      // enable continue button
-      SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm SD Reheat done... J18");
+    switch(HASH(msg)) {
+      case "Nozzle Parked"_hash: {
+        mediaPrintingState = AMPRINTSTATE_PAUSED;
+        mediaPauseState    = AMPAUSESTATE_PARKED;
+        // enable continue button
+        SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm SD print paused done... J18");
+        break;
+      }
+      case "Load Filament"_hash: {
+        mediaPrintingState = AMPRINTSTATE_PAUSED;
+        mediaPauseState    = AMPAUSESTATE_FILAMENT_OUT;
+        // enable continue button
+        SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm Filament is out... J18");
+        SENDLINE_DBG_PGM("J23", "TFT Serial Debug: UserConfirm Blocking filament prompt... J23");
+        break;
+      }
+      case "Filament Purging..."_hash: {
+        mediaPrintingState = AMPRINTSTATE_PAUSED;
+        mediaPauseState    = AMPAUSESTATE_PARKING;
+        // TODO: JBA I don't think J05 just disables the continue button, i think it injects a rogue M25. So taking this out
+        // disable continue button
+        // SENDLINE_DBG_PGM("J05", "TFT Serial Debug: UserConfirm SD Filament Purging... J05"); // J05 printing pause
+
+        // enable continue button
+        SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm Filament is purging... J18");
+        break;
+      }
+      case "HeaterTimeout"_hash: {
+        mediaPrintingState = AMPRINTSTATE_PAUSED;
+        mediaPauseState    = AMPAUSESTATE_HEATER_TIMEOUT;
+        // enable continue button
+        SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm SD Heater timeout... J18");
+        break;
+      }
+      case "Reheat finished."_hash: {
+        mediaPrintingState = AMPRINTSTATE_PAUSED;
+        mediaPauseState    = AMPAUSESTATE_REHEAT_FINISHED;
+        // enable continue button
+        SENDLINE_DBG_PGM("J18", "TFT Serial Debug: UserConfirm SD Reheat done... J18");
+        break;
+      }
```

Comment:
This display is also taking a bad approach and needs the same fix.

Reply:
Looks like we should do a separated PR to refactor those UI code... ?

Reply:
Where does it come from ?

Reply:
Various => `ExtUI::onUserConfirmRequired(STRING)` => `AnycubicTFTClass::OnUserConfirmRequired(STRING)`.

**Proposal:** Add a parameter to `onUserConfirmRequired` taking an `enum` that corresponds to each unique confirm message, so that the enum value can be compared instead of the string. Maybe the string pointer doesn't need to be passed, and it can call a function to get the string based on the enum. This function would then be used in other places where an enum could be passed instead of a string pointer.

Reply:
A small number of messages are ever sent to `onUserConfirmRequired_P` so it's quite doable:
```cpp
PSTR("Load Filament")
PSTR("Filament Purging...")
GET_TEXT(MSG_NOZZLE_PARKED)
GET_TEXT(MSG_HEATER_TIMEOUT)
PSTR("Reheat finished.")
PSTR("MMU2 Eject Recover")
PSTR("M43 Wait Called")
GET_TEXT(MSG_USERWAIT
GET_TEXT(MSG_PRINT_DONE)
PSTR("Delta Calibration in progress")
PSTR("Stow Probe")
```